manager/common/license.go

```diff
@@ -0,0 +1,62 @@
+package license
+import (
+	"log"
+	"sync"
+	"context"
+	rpc_license "github.com/enfabrica/enkit/manager/rpc"
+)
+
+type LicenseCounter struct {
+    mutex sync.Mutex
+    counter int
+    total int
+}
+
+func (c *LicenseCounter) increment(num int) int {
+    c.mutex.Lock()
+    if c.counter < c.total {
+        c.counter += num
+    }
+    c.mutex.Unlock()
+    return c.counter
+}
+
+func (c *LicenseCounter) decrement(num int) int {
+	c.mutex.Lock()
+	if c.counter - num >= 0 {
+		c.counter -= num
+	}
+	c.mutex.Unlock()
+	return c.counter
+}
+
+var totalLicenses = 3
+var licenseCounter = LicenseCounter{counter: 0, total: totalLicenses}
+type Server struct {
+}
+
+func (s *Server) Acquire(ctx context.Context, request *rpc_license.AcquireRequest) (*rpc_license.AcquireResponse, error) {
+	log.Printf("Received request for %d %s feature %s \n", request.Quantity, request.Vendor, request.Feature)
+	if licenseCounter.counter < licenseCounter.total {
+		log.Printf("Licenses in use: %d/%d \n", licenseCounter.increment(int(request.Quantity)), licenseCounter.total)
+		return &rpc_license.AcquireResponse{Available: true, Waiting: false, Missing: false}, nil
+	} else {
+		return &rpc_license.AcquireResponse{Available: false, Waiting: true, Missing: false}, nil
```

Comment:
The job ID should be returned to the client inside the response body so that if the client needs to poll for a license, the connection can be uniquely identified from other clients.
```
jobId := 0
...
jobId += 1
jobId = jobId % MAX_INT // prevent integer overflow
```

Reply:
See recommendations for changes to the interface below. I would actually recommend you return a random string, and just keep it in a hash table.

To generate the string, you can use something like [GenerateUID here](https://github.com/enfabrica/enkit/blob/master/astore/server/astore/astore.go#L55) (happy to move this in more of a library function).

A simple `map[string]*Client` for example, where client is something like:
```
type Client struct {
  IP string // extracted from grpc connection
  Owner string // ...
  Grabbed time.Time // when the license was grabbed
  Refreshed time.Time // when the license was last refreshed
  Interval time.Duration // max wait time for a refresh
}
```

... to "return licenses once the timer expires", and "refresh the timers", you can use something like (the pinger thread here)[https://github.com/enfabrica/enkit/blob/master/lib/icmp/pinger.go#L450], which runs a separate thread, which just waits for the next timer to expire, and invokes a callback. This should also be moved in a library, like `lib/goroutine`.